pil-analyzer/src/evaluator.rs

```diff
@@ -551,51 +516,219 @@ pub trait SymbolLookup<'a, T> {
+                        self.symbols
+                            .new_witness_column(name, SourceRef::unknown())?
+                    };
+                    self.local_vars.extend(
+                        Value::try_match_pattern(&value, &s.pattern).unwrap_or_else(|| {
+                            panic!("Irrefutable pattern did not match: {} = {value}", s.pattern)
+                        }),
+                    );
+                }
+                Operation::AddConstraint() => {
+                    let result = self.value_stack.pop().unwrap();
+                    self.symbols.add_constraints(result, SourceRef::unknown())?;
+                }
+            };
+        }
+        assert_eq!(self.value_stack.len(), 1);
+        Ok(self.value_stack.pop().unwrap())
+    }
+
+    fn expand(&mut self, expr: &'a Expression) -> Result<(), EvalError> {
+        match expr {
             Expression::Reference(reference) => {
-                evaluate_reference(reference, locals, type_args, symbols)?
+                let v = self.evaluate_reference(reference)?;
+                self.value_stack.push(v)
             }
-            Expression::PublicReference(name) => symbols.lookup_public_reference(name)?,
-            Expression::Number(n, ty) => evaluate_literal(n.clone(), ty, type_args)?,
-            Expression::String(s) => Value::String(s.clone()).into(),
-            Expression::Tuple(items) => Value::Tuple(
-                items
-                    .iter()
-                    .map(|e| evaluate(e, locals, type_args, symbols))
-                    .collect::<Result<_, _>>()?,
-            )
-            .into(),
-            Expression::ArrayLiteral(elements) => (Value::Array(
-                elements
-                    .items
-                    .iter()
-                    .map(|e| evaluate(e, locals, type_args, symbols))
-                    .collect::<Result<_, _>>()?,
-            ))
-            .into(),
-            Expression::BinaryOperation(left, op, right) => {
-                let left = evaluate(left, locals, type_args, symbols)?;
-                let right = evaluate(right, locals, type_args, symbols)?;
-                evaluate_binary_operation(&left, *op, &right)?
+            Expression::PublicReference(name) => self
+                .value_stack
+                .push(self.symbols.lookup_public_reference(name)?),
+            Expression::Number(n, ty) => {
+                self.value_stack
+                    .push(evaluate_literal(n.clone(), ty, &self.type_args)?)
+            }
+            Expression::String(s) => self.value_stack.push(Value::String(s.clone()).into()),
+            Expression::Tuple(_)
+            | Expression::ArrayLiteral(_)
+            | Expression::BinaryOperation(_, _, _)
+            | Expression::UnaryOperation(_, _)
+            | Expression::IndexAccess(_)
+            | Expression::FunctionCall(_) => {
+                let children_count = expr.children().count();
```

Comment:
While it looks neat to handle all those cases in a uniform way, it might be quite bad for performance.

Reply:
Indeed it is.